rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+Having multiple apps with some inconsistencies in how they are structured represents a difficulty for experienced developers as well as a high barrier to entry for new starters. Onboarding new team could be easier with fewer apps.
+
+### High maintenance burden
+Code dependencies are managed automatically through github with dependabot. Often a dependabot PR will be raised in each app, meaning that developers need to approve and merge that PR 8 times.
+
+Many dependabot PRs are now automatically merged, but each app is deployed separately. This means that changes that impact users can lead to brief periods of visual inconsistency. It also makes deploying breaking changes (particularly where `static` is involved) more difficult.
+
+Rails upgrades must also be repeated separately for each app. These upgrades can often involve a lot of work, particularly for major releases. This is especially problematic for applications that are less well maintained, have slightly different ways of doing things or are harder to test.
+
+### Deployments can be overly complicated
+Our current setup means that code changes often occur in one app and therefore several deployments must be carried out together to ensure that the user experience across the site remains consistent. This is difficult to achieve precisely because each application is different and can take a different amount of time to deploy.
+
+This will likely be an issue when deploying any upcoming changes to the GOV.UK brand.
+
+### Inconsistent testing approach
+The test suite is inconsistent across the rendering applications, which means that developers need to be familiar with more test suites than necessary, increasing the maintenance burden.
+
+Some apps use RSpec while others use Minitest. Some use Jasmine, some Cucumber and some Mocha. Many use the old controller tests, some have request tests, and some have both.
+
+Even when applications use the same testing frameworks there are still a lot of inconsistencies in how the tests are set up, for example in how fixture files are used, or how `content-store` and `search-api` are stubbed.
+
+### Duplicated assets and complex asset model
+Our assets (CSS and JavaScript files) are well optimised but having multiple apps means that some problems remain. If a user journey involves more than one app this means that duplicate page assets can be downloaded, reducing the speed of the site.
+
+For example if a user visits a page in one app that requires the button component CSS and then moves to a page on another app with the same component, that CSS file is downloaded once per page, rather than once.
+
+We aim to build our pages and assets to be as performant as possible, but repeated downloads invalidate some of that effort. This situation happens for any repeated components across the estate, where each app will re-deliver CSS and JavaScript to the user unnecessarily.
+
+### `router` / `router-api` are complicated and hard to maintain
```

Comment:
I'll add a section. The thought is that if we can get down to say 2 apps, e.g. `frontend` and `smartanswers` then `frontend` could just redirect routes to the `smartanswers` removing the need for `router` and `router-api`.

Reply:
Ah so do you mean have frontend act as proxy similar to how router is?

I imagine the ideal would be if we had a more conventional proxy routing system where we used a path prefix for apps other than frontend, so /chat to chat, /email to email-alert-frontend, /smart-answers for Smart Answers then everything else to frontend. Then we could get rid of the need for the complex system. Only thing that seems to be problematic is the existing URLs and how divisive it'd be to change them (with Smart Answers seeming easily most divisive there)

Reply:
Agree, if have still have separate frontend apps. A path prefix would make it super easy to hand routing off to the load balancer. (I know the content folk like to have easy human readable URLs, but there needs to be compromise between the value that brings users vs engineering complexity here). Dealing with existing URLs would be tricky, but not impossible.

Reply:
As a stopgap, could Frontend (instead of Router) read from Router API and route to the other apps? What additional load would that put on the system? We're pretty heavily cached by the CDN, right? What would be the effect on the end rendering time if a rails app did the routing rather than the go app? I imagine it would be vastly slower, but does vastly slower actually equate to much in terms of what the end user sees?

Reply:
I don't know for sure but I guess it'd be a humongous load difference given that router deals with all the 404s and then government-frontend gets the most app traffic. Whether we can cope with it though seems like something we could work out, particularly if we could applying auto-scaling given the app is pretty much stateless so should scale well. 

One thing I'm not clear on is whether Router conceals us a bit from some of the Ruby pains we have of only able to serve as many users as the number of threads we have per server, which can lead to requests in a backlog and eventually just aborted. But given Router already proxies to apps with that limitation I can't imagine it is that impactful.

What I'd probably suggest as a stop gap though is that rather than having Frontend start take Router API responsibilities you instead set Router to forward everything it doesn't know what to do with to Frontend, so Frontend can gradually take on handling error pages etc. You could then stop indexing in Router API for things to frontend so that the Router API database gradually gets smaller and smaller to be less impactful.

Reply:
I think we're _already_ doing that with government-frontend, but I may be wrong.

Reply:
I think that rather than dropping the router section, we should add another that says how far we towards solving the problems with this RFC.